### DISCUSSION TITLE
New API for Dart 3.0

### DISCUSSION THREAD

**Discussion by kuhnroyal** [cfug/dio#1753] [Ideas]
I was thinking a lot about this while playing with the wrong `stackTrace`s issue.
The whole concept of throwing a `DioError` in case of a server response with a bad status code seems wrong.

With the advent of sealed classes and pattern matching, it would be a nice time to design a new API for Dio.

Something along the lines:

```dart
sealed class Response {}

class Success extends Response {}
class Failure extends Response {}

class Call<T> {
  final RequestOptions request;

  Future<Response<T>> call();
}

final request = dio.get('/foo', failedRequestStatusCodes: [400...599]);
final response = await request().catchError((e, st) {
  print('some error');
});

switch (response) {
  case Success: print('success');
  case Failure: print('failure');
}
```

I think we can completely get rid of `DioError` with this. Response errors are just `Failure`s and don't need a stackTrace because they are not thrown around anymore. And all other errors don't need to be wrapped, instead we could provide some smart helpers for error handling.

What do you guys think? @AlexV525 @ueman

**Comment by ueman** [comment]
At that point you could also basically copy `http` API and let the user simply check the status code. I would rather keep the exceptions for bad status code and just not wrap any other exception.

Throwing for bad status code is one of the main differences to the `http` package.

**Reply by ueman** [reply]
I guess we should start by laying out our expectations for how dio deals with certain status codes/timeouts and other parts of the API before jumping straight to the solution.

(We should probably also write down those differences to the http package. I guess that will be very helpful for comparing various http clients)

**Reply by kuhnroyal** [reply]
I absolutely agree, I just want to get the ball rolling, it is not that far out :)
